Misc. Improvements

nbougalis opened the pull request.

scottschurr commented:
It looks like we're pretty consistently getting 30 unit test failures on several platforms with this code at the moment.  I didn't investigate further.

nbougalis commented:
Unit test failures are fixed; my bad.

HowardHinnant commented on src/ripple/basics/contract.h:
```diff
@@ -49,6 +60,8 @@ Throw (Args&&... args)
     static_assert (std::is_convertible<E*, std::exception*>::value,
         "Exception must derive from std::exception.");
+
+    LogThrow ("Throwing exception");
```
How about:
````
E e(std::forward<Args>(args)...);
LogThrow(std::string("Throwing exception: ") + e.what());
throw e;
````

HowardHinnant commented on src/ripple/basics/impl/contract.cpp:
```diff
@@ -39,6 +41,20 @@ accessViolation() noexcept
 
+void
+LogThrow (char const* title)
+{
```
If this takes a `std::string` instead we can build up a much richer message at the throw site.

HowardHinnant commented on src/ripple/app/main/Main.cpp:
```diff
@@ -470,6 +471,12 @@ int run (int argc, char** argv)
+            LogThrow ("Unhandled exception");
+            std::abort ();
+        });
+
```
On OS X this gives an inferior stack dump compared to what already happens.  With this change I get no output at the terminal with an unhandled exception, and the stack dump summary is "abort() called".  Without the change (and with the simulated unhandled exception I forced), I get:

libc++abi.dylib: terminating with uncaught exception of type ripple::SHAMapMissingNode: SHAMapMissingNode

both on the terminal, and in the stack dump.  This is both the type of the exception, and its what() message.

Guarding with:
````
#ifndef __APPLE__
````
is one possible solution.

nbougalis replied:
I just removed it - it's unhelpful.

scottschurr commented on src/ripple/protocol/impl/STVar.h:
```diff
@@ -54,12 +57,12 @@ class STVar
     STVar (STBase&& t)
     {
-        p_ = t.move(sizeof(d_), &d_);
+        p_ = t.move(max_size, &d_);
```
Hmmmm.  I wonder if using `max_size` here rather than `size_of(d_)` is a pessimization.  Since I was unacquainted with `aligned_storage` I ran an experiment in Wandbox.  Running gcc in Wandbox reports the actual size of `d_` is 80.  So, at least in gcc, we could be wasting 8 bytes of storage per `STVar`.  Clearly using `max_size` is also safe, so it's not really a bug, but it may be less efficient.

I'm not acquainted with `aligned_storage`.  I'm assuming the compiler will obey the alignment requirement if an `STVar` is created on the stack?

Just curious if anyone knows why `72` was picked.  If we do know, it would be good to leave a comment.

nbougalis replied:
It's not a pessimization in the sense that 72 is the largest fixed size object we allow in an `STVar`. The next larger one is 104 bytes. Perhaps a comment to that effect is in order.

Yes, `sizeof(d_)` gives us 80, but I tend to think of the extra 8 bytes as "padding". Maybe accessing them is OK (in fact, it _probably_ is) but I still think it's better to only use what we asked for.

scottschurr commented on src/ripple/protocol/impl/STVar.cpp:
```diff
@@ -61,37 +59,42 @@ STVar::STVar (STVar&& other)
 {
-    destroy();
-    if (rhs.on_heap())
+    if (&rhs != this)
```
I like this test, but I think technically it's not required.  My understanding is that, if `rhs` really is a temporary, `this` can never be the same as `&rhs`.  That being said, I like the safety of the check.  Assigning `STVar`s is not so time critical that we can't afford the check.

seelabs replied:
`a = std::move(a)` ?

nbougalis replied:
I wanted to avoid accidents along the lines of `x = std::move(x)` since they'd be catastrophic.

scottschurr commented on src/ripple/protocol/impl/STVar.cpp:
```diff
@@ -61,37 +59,42 @@ STVar::STVar (STVar&& other)
-    if (rhs.p_)
-        p_ = rhs.p_->copy(
-            sizeof(d_), &d_);
+    if (&rhs != this)
```
Good catch!

HowardHinnant commented on src/ripple/basics/contract.h:
```diff
@@ -33,11 +33,22 @@ namespace ripple {
 
+/** Generates and logs a call stack */
+void
+LogThrow (std::string title);
```
Should be `std::string const&`.

HowardHinnant commented:
👍 Just one minor comment.

scottschurr commented on src/ripple/basics/impl/Log.cpp:
```diff
@@ -340,21 +343,58 @@ Logs::format (std::string& output, std::string const& message,
+debugJournal()
 {
-    debugJournal_ = std::make_unique<beast::Journal>(sink);
+    return beast::Journal (dj().get());
```
This works, but I think it can be simplified to `return dj().get()`.  Just a thought.

scottschurr commented on src/ripple/basics/impl/contract.cpp:
```diff
@@ -39,6 +41,18 @@ accessViolation() noexcept
+            j << "    " << frame;
+        j << "----";
+    }
+}
```
Woah.  I'm sure this works, but it looks expensive to me since we're logging a backtrace of every `Throw` and `Rethrow` regardless of whether the `Throw` is unhanded at the top of the thread.

A cheaper solution (at runtime) might be to add similar logging in a `catch(...)` in the top function of every new thread.  That solution would also log unhanded exceptions coming from places like the standard library or soci, not just our own exceptions.  It would be cheaper because it would only log the unhanded exceptions -- not all of them.  Just my thoughts on the matter.

nbougalis replied:
We'll only log if `debugJournal().debug()` is enabled, which only happens if you run with `--debug` (at least, that's the intention)

scottschurr replied:
Yeah, but my understanding is that *all* our production servers are running at that log level.  So it's only cheaper if it's not a production server?

nbougalis replied:
Not debug log level - with the `--debug` command line argument, which activates the `debugJournal`. Otherwise, we get back a journal tied to a null sink, which should be at "disabled" level, and thus we never enter the `if`

scottschurr replied:
Oh.  Thanks.  I hadn't understood that distinction.  I thought `--debug` affected the log level.  Thanks for straightening me out.

Still.  The point of this exercise is to find out why our production servers are crashing.  So we'll want to enable this logging on our production servers (except our validators, which aren't crashing), right?  So in order for this change to be useful our production severs must pay the price of logging a stack trace for every `Throw` and `Rethrow`.

I'm not saying don't do it.  I'm just saying it's compute expensive and there is a less compute expensive and more thorough solution (since it catches all exceptions).  But that less compute expensive solution requires more coding work than this one requires.

To be clear, I'm okay with what you're doing.  I'm just presenting what I suspect is a better alternative.

nbougalis replied:
I have another plan for the unhandled exception - it involves `__cxa_throw` and some interesting features of `ld` - it's a bit involved, so will likely need to do it over the weekend.

I'm fine with what you're suggesting by the way – in fact, I think it's a great idea for all the threads under our control.

HowardHinnant replied:
I've already discussed this with Nik, but I'm repeating here to get the word out to a wider audience:  On OS X we don't want to catch, log and rethrow at the top of the thread as this destroys information on OS X.  We get a better stack dump without the catch/rethrow.  Unfortunately gcc operates differently, with a lousy stack dump whatever we do.  `#ifndef __APPLE__` can be used to guard such changes.

scottschurr replied:
Thanks @HowardHinnant.  Is that because Apple implemented a much better equivalent to gcc's `execute_native_thread_routine(void* __p)`?  https://chromium.googlesource.com/chromiumos/third_party/gcc/+/5cfcf8452e8f755d2b2899d96f93afdb4672de16/gcc/gcc-4.6.0/libstdc++-v3/src/c++11/thread.cc

HowardHinnant replied:
Yes, here is the libc++ `execute_native_thread_routine` equivalent:
https://github.com/llvm-mirror/libcxx/blob/master/include/thread#L343-L379
The crucial difference is that the libc++ version has no top level `try/catch`.  Unhandled exceptions are literally unhandled and another library (the unwind library) is responsible for calling `terminate` in this circumstance.  When it does, it does so without having unwound the stack, and thus one gets a stack trace from the point of `throw`.

I can think of no technical reason why gcc could not do this as well, because both platforms follow the same Itanium ABI specification for exception throwing.

scottschurr replied:
Thanks.  I think I'm catching on.  The problem is that once the `catch` occurs the stack below is lost.  We can't see where the `throw` came from.  All we can log is that we are where we are and we have this $%^& exception in out hands.

HowardHinnant replied:
You got it perfectly.

scottschurr commented:
:+1: It all looks good to me.  The `Throw` logging looks compute expensive to me, but it looks like it will work.

nbougalis commented:
Thanks @scottschurr. I'll make some changes to address a couple of issues identified in the review.

nbougalis commented:
@scottschurr, @HowardHinnant: I've addressed your comments. I changed the second commit "Serialize access to the debug journal" and removed the printing of the call stack from `LogThrow` - we just print a short header message now.

scottschurr commented:
Looks great to me. :+1:

nbougalis commented:
Rebased to b13 - added Passed label.

nbougalis commented:
Merged as acebbf58eb9d09e8763b485623607bed62a733d2, a039e7593a33ad13b0bc4ccd2df4f4d7eb87a3c2 and 97c89168f75af89cc09131fdba03d8fb0decd1b9
